python/search_completer/hierarchical_search_completer.py

```diff
@@ -0,0 +1,227 @@
+        """
+        return self._show_entities_only
+
+    @show_entities_only.setter
+    def show_entities_only(self, is_set):
+        """
+        See getter documentation.
+        """
+        self._show_entities_only = is_set
+
+    def _set_item_delegate(self, popup, text):
+        """
+        Sets the item delegate for the popup widget.
+
+        :param popup: Qt Popup widget receiving the delegate.
+        :paarm text: Text from the current search.
+        """
+        # deferred import to help documentation generation.
+        from .hierarchical_search_result_delegate import HierarchicalSearchResultDelegate
+        self._delegate = HierarchicalSearchResultDelegate(popup, text)
+        popup.setItemDelegate(self._delegate)
+
+    def _launch_sg_search(self, text):
+        """
+        Launches a search on the Shotgun server.
+
+        :param str text: Text to search for.
+
+        :returns: The :class:`~tk-framework-shotgunutils:shotgun_data.ShotgunDataRetriever`'s job id.
+        """
+        # FIXME: Ideally we would use the nav_search_entity endpoint to compute the path to the root.
+        # Unfortunately there is a bug a the moment that prevents this.
+        if not self._search_root:
+            root_path = "/"
+        else:
+            root_path = "/Project/%d" % self._search_root.get("id")
+
+        return self._sg_data_retriever.execute_nav_search_string(
+            root_path, text
+        )
+
+    def _handle_search_results(self, data):
+        """
+        Populates the model associated with the completer with the data coming back from Shotgun.
+
+        :param dict data: Data received back from the job sent to the
+            :class:`~tk-framework-shotgunutils:shotgun_data.ShotgunDataRetriever` in :method:``_launch_sg_search``.
+        """
+        matches = data["sg"]
+
+        # When showing only entities, skip entries that aren't.
+        if self._show_entities_only:
+            matches = filter(
+                lambda x: x["ref"]["type"] is not None and x["ref"]["id"] is not None,
+                matches
+            )
+
+        if len(matches) == 0:
+            item = QtGui.QStandardItem("No matches found!")
+            item.setData(self.MODE_NOT_FOUND, self.MODE_ROLE)
+            self.model().appendRow(item)
+
+        # Payload looks like:
+        # [
+        #     {
+        #         "label": "bunny_020",
+        #         "incremental_path": [
+        #             "/Project/65",
+        #             "/Project/65/Shot",
+        #             "/Project/65/Shot/sg_sequence/Sequence/5"
+        #         ],
+        #         "path_label": "Shots",
+        #         "ref": {
+        #             "id": 5,
+        #             "type": "Sequence"
+        #         },
+        #         "project_id": 65
+        #     },
+        #     ...
+        # ]
+
+        # insert new data into model
+        for d in matches:
```

Comment:
can we change this to `for data in data_matches` or something a little more descriptive?